sunpy/map/mapbase.py

```diff
@@ -1592,6 +1587,56 @@ def submap(self, bottom_left, *, top_right=None, width: (u.deg, u.pix) = None, h
         new_map = self._new_instance(new_data, new_meta, self.plot_settings)
         return new_map
 
+    @seconddispatch
+    def _parse_submap_input(self, bottom_left, top_right, width, height):
+        """
+        Should take any valid input to submap() and return bottom_left and
+        top_right in pixel coordinates.
+        """
+        pass
+
+    @_parse_submap_input.register(u.Quantity)
+    def _parse_submap_quantity_input(self, bottom_left, top_right, width, height):
+        if bottom_left.shape != (2, ):
+            raise ValueError('bottom_left must have shape (2, ) when specified as a Quantity')
+        if top_right is not None:
+            if top_right.shape != (2, ):
+                raise ValueError('top_right must have shape (2, ) when specified as a Quantity')
+            if not top_right.unit.is_equivalent(u.pix):
+                raise TypeError("When bottom_left is a Quantity, top_right "
+                                "must be a Quantity in units of pixels.")
+            # Have bottom_left and top_right in pixels already, so no need to do
+            # anything else
+        else:
+            if not (width.unit.is_equivalent(u.pix) and
+                    height.unit.is_equivalent(u.pix)):
+                raise TypeError("When bottom_left is a Quantity, width and height "
+                                "must be a Quantity in units of pixels.")
+            # Add width and height to get top_right
+            top_right = u.Quantity([bottom_left[0] + width, bottom_left[1] + height])
+        return bottom_left, top_right
+
+    @_parse_submap_input.register(SkyCoord)
+    def _parse_submap_coord_input(self, bottom_left, top_right, width, height):
+        if top_right is not None:
+            if not isinstance(top_right, SkyCoord):
+                raise TypeError("When bottom_left is a SkyCoord, top_right "
+                                "must also be a SkyCoord.")
+        else:
+            if not (width.unit.is_equivalent(u.deg) and
+                    height.unit.is_equivalent(u.deg)):
+                raise TypeError("When bottom_left is a SkyCoord, width and height "
+                                "must both be a Quantity in units convertable to degrees.")
```

Comment:
Isn't this check also done in `get_rectangle_coordinates`? I would have thought that all the validation we need for this branch is in the helper, so can't we just call the helper and convert to pixels in this function?